### DISCUSSION TITLE
Would it be possible to add "Find Usages" and "Smart Rename" to built-in editor / GDScript?

### DISCUSSION THREAD

**Discussion by daphMonk** [godotengine/godot-proposals#7952] [Editor]
Find usages or Find all references, returns a search result with all the places where the variable (or function) is being used. Smart Rename allows to rename a variable, but not based on the name being the same, but the variable having the same context, ie : Smart Rename would allow to modify all the references found for a given variable through find all references.

I am aware that due to GDScript being a dynamic language, this is a feature that is hard to implement. But I wonder if in the very least, this couldn't be a feature that could be implemented for the people who use strong type in GDScript. Or, it could return the results that it's able to find, fully stating that there might be missing references through dynamic language. These two functionalities could even be optional and need to be enabled in editor settings.

As this annoys me greatly, I'd be willing to work on this feature if someone were to point me in the right direction. (If I can see how ctrl + click is implemented, and probably how ctrl + shift + f is as well, I believe I could find a way to do it in reverse. Of course, I can't guarantee results before I look into it but I'd be willing to try.)

**Comment by Calinou** [comment]
Thanks for your interest in contributing this feature :slightly_smiling_face: 

<kbd>Ctrl + Click</kbd> functionality is known as *symbol lookup* (there is a "Lookup Symbol" shortcut to trigger it from the keyboard). You can find its implementation here: https://github.com/godotengine/godot/blob/0ca8542329888e8dccba89d59d3b728090c29991/editor/plugins/script_text_editor.cpp#L882

As explained in https://github.com/godotengine/godot-proposals/issues/899, refactor tooling (i.e. symbol renaming) is already present in the GDScript LSP server, but it's not exposed in the built-in script editor.

**Reply by daphMonk** [reply]
Thanks! I'll look into both of these places. :)
